modules/sov-ibc/src/rpc/helpers.rs

```diff
@@ -40,3 +43,67 @@ pub fn to_jsonrpsee_error(err: impl ToString) -> ErrorObjectOwned {
         None::<String>,
     )
 }
+
+/// Trait for proof agnostic storage value retrieval.
+pub trait StorageValue<V> {
```

Comment:
I introduced `StorageValue` trait along with concrete implementations for `WithoutProof` and `WithProof`.

Thus, I avoided having duplicated retrievals with and without proofs.

Reply:
Liked the idea. Neat!

To take it a step further, see if something as below is applicable? This should be implemented on IbcContext. (Relevant to the question you asked here)

```rust
pub trait StorageValue<V> {
where
    S: Spec,
{

    fn value_at_key<K, C>(
        &self,
        key: &K,
        state_map: &StateMap<K, V, C>,
        include_proof: IncludeProof,
    ) -> RpcResult<V>
    where
        C: StateCodec,
        <C as StateCodec>::ValueCodec: StateItemCodec<V>,
        <C as StateCodec>::KeyCodec: StateItemCodec<K>,
    {
        
        match include_proof {
            IncludeProof::NoProof => {
                let value = state_map.get(key, *self.working_set.borrow_mut());
                Ok(value)
            }
            IncludeProof::WithProof => {
                let result = state_map.get_with_proof(key, *self.working_set.borrow_mut());
                let value = result
                    .value
                    .map(|bytes| state_map.codec().value_codec().decode_unwrap(bytes.value()))
                    .ok_or_else(|| to_jsonrpsee_error("value not found"))?;
                Ok(value)
            }
            
        }
    }
    }
}
```

Reply:
https://github.com/informalsystems/sovereign-ibc/pull/123#discussion_r1556138172

Reply:
I didn't use an enum because, in that case, the choice would be handled at runtime. The type generics would handle it in compile time.

Reply:
I think that it wouldn't be a big improvement if I implemented these traits directly on `IbcContext`.

A simple analogy would be serialization.

```rs
trait Serializer {
  fn serialize<T>:(value: &T) where T: serde::Serializable -> Vec<u8>;
}

impl Serializer for Json {
  fn serialize<T>:(value: &T) where T: serde::Serializable-> Vec<u8> { ... }
}

impl Serializer for Yaml {
  fn serialize<T>:(value: &T) where T: serde::Serializable -> Vec<u8> { ... }
}

#[derive(serde::Serialize)]
struct MyStruct;

impl MyStruct {
  fn serialized<S: Serializer>(&self) -> Vec<u8> {
    S::serialize(&self)
  }

  fn some_serialized_value<S: Serialize>(&self) -> Vec<u8> {
    self.do_some_updates().serialized()
  }
}
```

My current implementation directly calls `S::serialize`, avoiding the `MyStruct::serialized`.


```rs
impl MyStruct {
  fn some_serialized_value<S: Serialize>(&self) -> Vec<u8> {
    S::serialize(self.do_some_updates())
  }
}
```

Reply:
That's fine. The current implementation is clean enough as it is. Just two more points:

- The generic `V` in `WithProof<V>` and `WithoutProof<V>` seem redundant. I think it can be removed.
- On a broader note, the `WithoutProof` case will likely remain unused in practice since users can't specify whether proof should be included with their request or not. Though, I'm okay with keeping that, but would be useful to add a comment explaining why it's there, in case a relevant feature is requested later.

Reply:
- This is a very good question. I need it to specify the return type in the `StorageValue::Output`. I am looking into if I can remove this using GAT.
- Makes sense. I will add an explanation.

Reply:
refactor with GAT 0fc825a :smile: